WWST-7794, Add a DTH code for ABL CPX Lighting(Light with a motion sensor)

RaihaPark opened the pull request:
For the ABL CPX Lighting, we create a folder "zigbee-motion-sensor-light-light.src" in the smartthings directory.
In the test with local public, the device works fine.
It contains 3 files : README.md , led-fan-cpx-light.groovy , led-cpx-motion-sensor-child.groovy

It makes two device cards : lightings and motion sensor

In the lightings card, you can control on / off, brightness level and color temperature.
In the motion sensor card, you can see the sensor works and can create routines.

Please review and approve this request.

RaihaPark committed:
Create led-cpx-light.groovy

First commit

RaihaPark committed:
Create led-cpx-light.groovy

First commit

RaihaPark committed:
Update led-cpx-light.groovy

Line aligning

RaihaPark committed:
Delete led-cpx-light.groovy

Move to other folder

RaihaPark committed:
Create README.md

First commit

RaihaPark committed:
Create led-cpx-motion-sensor-child.groovy

First commit

RaihaPark committed:
Update led-cpx-motion-sensor-child.groovy

Line aligning

RaihaPark committed:
Update README.md

Add definition of motion sensor capability

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+/**
+ *  Copyright 2017 SmartThings
```
date

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+import groovy.json.JsonOutput
+
+metadata {
+	definition(name: "LED CPX light", namespace: "SAMSUNG LED", author: "SAMSUNG LED", runLocally: true, minHubCoreVersion: '000.019.00012', executeCommandsLocally: true, genericHandler: "Zigbee") {
```
These flags `runLocally: true, minHubCoreVersion: '000.019.00012', executeCommandsLocally: true, genericHandler: "Zigbee"` shouldn't be here.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+		capability "Refresh"
+		capability "Switch"
+		capability "Switch Level"
+		capability "Light"
```
This capability is deprecated.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+		capability "Switch Level"
+		capability "Light"
+		
+		attribute "colorName", "string"
```
This won't do anything in the (not so) new app. Declaring attributes this way worked the Classic app, which is not supported for some time now.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+			}
+			sendEvent(event)
+		}
+	} else if (zigbeeMap.cluster == "0406" && zigbeeMap.attrId == "0000") {
```
Can you change those "magic numbers" into constants, like it's done in another DTHes?

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+		
+	sendEvent(name: "colorName", value: genericName)
+	}
+}
```
This won't have any effect in the SmartThings app. If it's crucial, you'd need to create your own custom capability.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+	return
+}
+
+def addChildSensor(){
```
nit: space after round brackets

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+		it.device.deviceNetworkId == "${device.deviceNetworkId}:1"
+	}
+	
+	if(childDevice != null) {
```
nit: space after `if`

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-motion-sensor-child.groovy:
```diff
@@ -0,0 +1,53 @@
+/**
+ *  Copyright 2017 SmartThings
```
date

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-motion-sensor-child.groovy:
```diff
@@ -0,0 +1,53 @@
+}
+
+def configure() {
+	return zigbee.configureReporting(0x406, 0x0000, 0x18, 30, 600, null)
```
This won't send any Zigbee messages, as child device isn't directly "tied" with a device. But since the same message is returned from parent's `configure()` method, it can be easily deleted from here...

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-motion-sensor-child.groovy:
```diff
@@ -0,0 +1,53 @@
+metadata {
+	definition (name: "ITM CPX Motion sensor child", namespace: "SAMSUNG LED", author: "SAMSUNG LED", ocfDeviceType: "x.com.st.d.sensor.motion") {
+		capability "Motion Sensor"
+		capability "Configuration"
```
...alongside with `Configuration` capability.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-motion-sensor-child.groovy:
```diff
@@ -0,0 +1,53 @@
+		capability "Motion Sensor"
+		capability "Configuration"
+		capability "Refresh"
+		capability "Health Check"
```
`Health Check` capability, without `ping()` method is pointless. Either remove capability from the child DTH, or include `ping()` method. I'd personally go with 1st option and add a zigbee message to the parent's `ping()` that would bring child back from being offline.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+		zigbee.colorTemperatureRefresh() +
+		zigbee.onOffConfig(0, 300) +
+		zigbee.levelConfig()
+}
```
any `config` messages aren't a good fit for a `refresh()` method.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,187 @@
+	def zigbeeMap = zigbee.parseDescriptionAsMap(description)
+	if (event) {
+		if (event.name == "level" && event.value == 0) {
+		} else {
```
==
```
if (!(event.name == "level" && event.value == 0)) {
```

PKacprowiczS reviewed (commented):
There's also an error in the DTH's path.
`devicetypes/smartthings/zigbee-motion-sensor-light-light.src/led-cpx-light.groovy`
(one `light` too many)

RaihaPark committed:
Edit codes

Edit codes

RaihaPark committed:
Edit codes

Edit codes

RaihaPark committed:
Edit codes

Edit codes

RaihaPark committed:
Update led-cpx-light.groovy

Delete configuration in the refresh function.
Delete color temperature naming function.

RaihaPark commented:
Thank you for reviewing. I fixed what you said :).

RaihaPark committed:
Update led-cpx-light.groovy

Edited conditional statement at line 73 in the parse function.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -13,24 +13,18 @@
-		capability "Refresh"
-		capability "Switch"
-		capability "Switch Level"
-		capability "Light"
```
I was only referring `Light` capability.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -79,7 +77,8 @@ def parse(String description) {
 		}
-	} else if (zigbeeMap.cluster == "0406" && zigbeeMap.attrId == "0000") {
+	} else if (zigbeeMap.cluster == MOTION_CLUSTER_VALUE 
+		   && zigbeeMap.attrId == MOTION_STATUS_VALUE) {
```
Line break isn't necessary here.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -88,7 +87,8 @@ def parse(String description) {
 		def cluster = zigbee.parse(description)
-		if (cluster && cluster.clusterId == 0x0006 && cluster.command == 0x07) {
+		if (cluster && cluster.clusterId == ON_OFF_CLUSTER_VALUE
+		    && cluster.command == 0x07) {
```
Same here.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,167 @@
+	if (event) {
+		if (!(event.name == "level" && event.value == 0)) {
+			if (event.name == "colorTemperature") {
+				setGenericName(event.value)
```
You're calling removed function here.

RaihaPark committed:
Update led-cpx-light.groovy

Aligning codes.

RaihaPark commented:
Thank you very much :). I have fixed those.

RaihaPark committed:
Update led-cpx-light.groovy

Edit try-catch in line 151

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,166 @@
+		zigbee.levelRefresh() +
+		zigbee.colorTemperatureRefresh() +
+		zigbee.onOffConfig(0, 300) +
+		zigbee.levelConfig()
```
You've missed those unwanted config cmds in refresh()

greens commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,166 @@
+	
+	def childDevice = getChildDevices()?.find {
+		it.device.deviceNetworkId == "${device.deviceNetworkId}:1"
+	}
```
you can also just use the object returned by `addChildDevice` instead of calling this find immediately after

PKacprowiczS replied:
Bump. 
```
def childDevice = addChildDevice("ITM CPX Motion sensor child", dni, device.hub.id, [completedSetup: true, label: "${componentLabel}", isComponent: false])
childDevice.sendEvent(name: "motion", value: "inactive")
```

greens commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-motion-sensor-child.groovy:
```diff
@@ -0,0 +1,47 @@
+ */		
+ 						
+import physicalgraph.zigbee.clusters.iaszone.ZoneStatus						
+import physicalgraph.zigbee.zcl.DataType						
```
these seem unused

greens commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,166 @@
+ *
+ *  LED CPX light
+ *
+ *  Author: SAMSUMG LED
```
typo?

greens reviewed (commented):
Why not just include motion detection on the main DTH? Why break it out into a child?

RaihaPark committed:
Update led-cpx-light.groovy

Edited and aligned codes.

RaihaPark committed:
Update led-cpx-motion-sensor-child.groovy

Edited and aligned codes.

RaihaPark committed:
Update README.md

Edited contents.

RaihaPark commented:
Hello @PKacprowiczS @greens ,

I edited and aligned codes.
When I test with codes there is some problems that no initialized and no status updated.
So I have had fixed.

One of reason is a magic number.
The format of "0406", "0000" in the  line @83 is different from that in the line @108 and @123.
So if I use magic number, there are format difference occurred. Then, the sensor status doesn't update.

Plus, line @23 is a problem.
When I deleted codes from "runLocally" to end, there is a parser problem occurred.
Parser works with many errors.
It sends end event to the hub which contains "0B00" when I control on and off.
So I turn on the lighting but the button of plug-in shows off.
And doesn't initial status update also.
Furthermore periodically(I think it is health check interval) sends "off" to the hub, so the button of plug-in change to off even the lighing is on.

By those reasons I think I can't reduce codes in line @23 and use magic number.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-motion-sensor-child.groovy:
```diff
@@ -12,36 +12,28 @@
+		standardTile("refresh", "device.refresh", inactiveLabel: false, decoration: "flat", width: 2, height: 2) {				
+			state "default", action: "refresh.refresh", icon: "st.secondary.refresh"			
+		}				
+		main(["motion"])				
```
Adding those extra spaces isn't necessary, right?

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -46,49 +50,44 @@ metadata {
 
-private getMOTION_CLUSTER_VALUE() { 0x0406 }
-private getMOTION_STATUS_VALUE() { 0x0000 }
-private getON_OFF_CLUSTER_VALUE() { 0x0006 }
```
Don't delete those.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -46,49 +50,44 @@ metadata {
-		childDevice.sendEvent(zigbeeMap)
 	} else {
 		def cluster = zigbee.parse(description)
-		if (cluster && cluster.clusterId == ON_OFF_CLUSTER_VALUE && cluster.command == 0x07) {
```
You still are using `int` here, so I don't understand coming back to a plain number instead of a constant.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -46,49 +50,44 @@ metadata {
+		} else {
+			def zigbeeMap = zigbee.parseDescriptionAsMap(description)
+			
+			if (zigbeeMap.cluster ==  "0406" && zigbeeMap.attrId == "0000") {
```
you can use, respectively, `zigbeeMap.clusterInt` and `zigbeeMap.attrInt` and use previously defined constants, instead of coming back to the mediocre solution

RaihaPark replied:
Great! It works well. Thank you very much.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -101,66 +100,64 @@ def on() {
-		zigbee.levelConfig()
-}
-
-def configure() {
```
Why did you remove `configure()` function?

RaihaPark replied:
Oh, I didn't remove that. I just cleaned up configure() and refresh() as you mentioned before. I separated refresh and configure commands.

PKacprowiczS replied:
Oh yeah, you're right. Sorry, missed that.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -101,66 +100,64 @@ def on() {
-
-def configure() {
-	def cmds = delayBetween([
-		sendEvent(name: "checkInterval", value: 2 * 10 * 60 + 1 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID]),
```
Side note, `sendEvent` shouldn't be here, as it isn't a hub action.

RaihaPark replied:
Yes, you're right. And sometimes it occurs errors. So I deleted that.

RaihaPark committed:
Update led-cpx-motion-sensor-child.groovy

Removed unnecessary spaces.

RaihaPark committed:
Update led-cpx-light.groovy

Added magic numbers and edited codes.

RaihaPark committed:
Update led-cpx-light.groovy

Aligned codes.

RaihaPark committed:
Update led-cpx-light.groovy

Cleaned up codes.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,166 @@
+private getMOTION_CLUSTER_VALUE() { 0x0406 }
+private getMOTION_STATUS_VALUE() { 0x0000 }
+private getON_OFF_CLUSTER_VALUE() { 0x0006 }
+private getON_OFF_COMMAND_VALUE() { 0x07 }
```
`ON_OFF_COMMAND_VALUE` isn't an accurate name here. `0x07` command, as it is global command, stands for _Configure Reporting Response_ and it is basically a confirmation that device received and acknowledged previously send `zigbee.levelConfig()`. If you're up to a heavy-weight read, you can find out more about it in the [ZCL Reference Document](https://zigbeealliance.org/wp-content/uploads/2019/12/07-5123-06-zigbee-cluster-library-specification.pdf), chapter 2.5.8

PKacprowiczS replied:
Generally, I'd drop `_VALUE` from each of those getters.
And I'd name one in line 68 more like `getMOTION_STATUS_ATTRIBUTE`.

RaihaPark replied:
I changed their name as like zigbee standard. Thank you very much.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,166 @@
+	
+	if(childDevice != null) {
+		childDevice.sendEvent(name: "motion", value: "inactive")
+	}
```
Here `childDevice` will be definitely null, as it isn't even declared. You need to declare it before `try {} catch {}` block, and then assign to it returned value from `addChildDevice`, as discussed earlier.
Also, missing space after `if`.

RaihaPark replied:
I wonder if I edited that correctly.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,166 @@
+import groovy.json.JsonOutput
+
+metadata {
+	definition(name: "LED CPX light", namespace: "SAMSUNG LED", author: "SAMSUNG LED", runLocally: true, minHubCoreVersion: '000.019.00012', executeCommandsLocally: true, genericHandler: "Zigbee") {
```
Those flags after `author` really shouldn't be here. If you're witnessing errors in IDE, they're caused by something else.

RaihaPark replied:
I think that is characteristic of parent-child device. If I deleted those, device works very confused. I attached logs when I send on/off control after deleted.

3:37:37: debug description is catchall: 0104 0006 01 01 0000 00 FC19 00 00 0000 0B 01 0B00
3:37:37: debug description is catchall: 0104 0006 01 01 0000 00 FC19 00 00 0000 0B 01 0100
3:37:37: debug description is on/off: 1
3:37:31: debug description is catchall: 0104 0006 01 01 0000 00 FC19 00 00 0000 0B 01 0B00
3:37:30: debug description is catchall: 0104 0006 01 01 0000 00 FC19 00 00 0000 0B 01 0000
3:37:30: debug description is on/off: 0

From bottom to top 3 lines are "off" and 3 lines are "on". The location of log is just before line @73. They are came in to the parser. The parser send each of them as an event. Then plug-in works incorrectly "off" "off" "off" "on" "on" "off" but device works correctly "off" "on".

PKacprowiczS replied:
Can you comment out line 70 and see what happens?

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,166 @@
+ */
+
+import physicalgraph.zigbee.zcl.DataType
+import groovy.json.JsonOutput
```
Still unused.

RaihaPark replied:
Oh, I see. I have deleted.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,166 @@
+}
+
+def updated() {
+	refresh()
```
Perhaps an "emergency childDevice creation" would be a better use of `updated()` than calling `refresh()`.
```
if (!childDevices) {
	addChildSensor()
}
```

RaihaPark replied:
OK, I changed that.

RaihaPark committed:
Update led-cpx-light.groovy

Edited codes.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -79,16 +76,16 @@ def parse(String description) {
-			if (zigbeeMap.clusterInt ==  MOTION_CLUSTER_VALUE && zigbeeMap.attrInt == MOTION_STATUS_VALUE) {
+			if (zigbeeMap.clusterInt ==  MOTION_CLUSTER && zigbeeMap.attrInt == MOTION_STATUS_ATTRIBUTE) {
 				def childDevice = getChildDevices()?.find {
-					it.device.deviceNetworkId == "${device.deviceNetworkId}:1" 
```
This was correct way of finding already created child device.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -153,14 +152,16 @@ def addChildSensor() {
+	if (childDevice != null) {
+		childDevice.sendEvent(name: "motion", value: "inactive")
+	}
+	
```
you're still referring to a `null` value...

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -79,16 +76,16 @@ def parse(String description) {
+			if (zigbeeMap.clusterInt ==  MOTION_CLUSTER && zigbeeMap.attrInt == MOTION_STATUS_ATTRIBUTE) {
 				def childDevice = getChildDevices()?.find {
-					it.device.deviceNetworkId == "${device.deviceNetworkId}:1" 
+					it.device.deviceNetworkId == addChildSensor() 
```
In such case, almost every incoming message would try to create a new child device.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -153,14 +152,16 @@ def addChildSensor() {
 		addChildDevice("ITM CPX Motion sensor child", dni, device.hub.id, [completedSetup: true, label: "${componentLabel}", isComponent: false])
 	} catch (e) {
 		log.warn "Failed to add ITM Fan Controller - $e"
 	}
```
```
def childDevice

try {
	String dni = "${device.deviceNetworkId}:1"
	childDevice = addChildDevice("ITM CPX Motion sensor child", dni, device.hub.id, [completedSetup: true, label: "${componentLabel}", isComponent: false])
} catch (e) {
	log.warn "Failed to add ITM Fan Controller - $e"
}

if (childDevice != null) {
	childDevice.sendEvent(name: "motion", value: "inactive")
}
```

RaihaPark replied:
Oh, I see. I fixed that.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -153,14 +152,16 @@ def addChildSensor() {
-	if(childDevice != null) {
-		childDevice.sendEvent(name: "motion", value: "inactive")
-	}
+	return "${device.deviceNetworkId}:1"
```
This shouldn't be returned.

RaihaPark committed:
Update led-cpx-light.groovy

Edited codes

RaihaPark commented:
@PKacprowiczS 

Thanks for your review. Here is logs :)
Those logs are "off" and "on" same to before.

6:26:12: debug [raw:0104 0006 01 01 0000 00 FC19 00 00 0000 0B 01 0B00, profileId:0104, clusterId:0006, sourceEndpoint:01, destinationEndpoint:01, options:0000, messageType:00, dni:FC19, isClusterSpecific:false, isManufacturerSpecific:false, manufacturerId:0000, command:0B, direction:01, data:[0B, 00], clusterInt:6, commandInt:11]
6:26:12: debug [raw:0104 0006 01 01 0000 00 FC19 00 00 0000 0B 01 0100, profileId:0104, clusterId:0006, sourceEndpoint:01, destinationEndpoint:01, options:0000, messageType:00, dni:FC19, isClusterSpecific:false, isManufacturerSpecific:false, manufacturerId:0000, command:0B, direction:01, data:[01, 00], clusterInt:6, commandInt:11]
6:26:12: debug [:]
6:26:09: debug [raw:0104 0006 01 01 0000 00 FC19 00 00 0000 0B 01 0B00, profileId:0104, clusterId:0006, sourceEndpoint:01, destinationEndpoint:01, options:0000, messageType:00, dni:FC19, isClusterSpecific:false, isManufacturerSpecific:false, manufacturerId:0000, command:0B, direction:01, data:[0B, 00], clusterInt:6, commandInt:11]
6:26:09: debug [raw:0104 0006 01 01 0000 00 FC19 00 00 0000 0B 01 0000, profileId:0104, clusterId:0006, sourceEndpoint:01, destinationEndpoint:01, options:0000, messageType:00, dni:FC19, isClusterSpecific:false, isManufacturerSpecific:false, manufacturerId:0000, command:0B, direction:01, data:[00, 00], clusterInt:6, commandInt:11]
6:26:09: debug [:]

PKacprowiczS commented:
Are those logs with commented out 70th line, `def event = zigbee.getEvent(description)`, as I've suggested?

RaihaPark commented:
Yes, I changed the codes to this.


67 : def parse(String description) {
68 :  	def event = zigbee.getEvent(description)
69 :  	def zigbeeMap = zigbee.parseDescriptionAsMap(description)
70 :  	log.debug "$zigbeeMap"

RaihaPark committed:
Update zigbee-rgbw-bulb.groovy

Changed the "deviceJoinName" in line @43 from "Juno Connect" to "RetroBasics RGBW" by the request of customer.

PKacprowiczS commented:
Just to be on the same page, by saying "comment out line 70 (well, in your recent example 68), `def event = zigbee.getEvent(description)`", I literally mean changing this line to `// def event = zigbee.getEvent(description)`.

RaihaPark commented:
If I change line 70 to "// def event = zigbee.getEvent(description)", the device doesn't controllable and an error message pops that "A network or server error occurred. Try again later.".
And I tested other DTH code that "zigbee-rgbw-bulb.groovy".
I deleted  after author.
It showed same result as like this project.
I attached logs as follows.

5:29:47: debug description is catchall: 0104 0006 08 01 0000 00 3361 00 00 0000 0B 01 0B00
5:29:47: debug description is catchall: 0104 0006 08 01 0000 00 3361 00 00 0000 0B 01 0B00
5:29:47: debug description is on/off: 1
5:29:02: debug description is catchall: 0104 0006 08 01 0000 00 3361 00 00 0000 0B 01 0B00
 5:29:02: debug description is on/off: 1
5:28:17: debug description is catchall: 0104 0006 08 01 0000 00 3361 00 00 0000 0B 01 0B00
5:28:17: debug description is catchall: 0104 0006 08 01 0000 00 3361 00 00 0000 0B 01 0B00
5:28:17: debug description is on/off: 1
5:28:17: debug description is on/off: 1

RaihaPark commented:
Hello @PKacprowiczS, is everything O.K? I wonder, if there is some problems.

PKacprowiczS commented:
Yeah, sorry about the radio silence. I gotta say, I have trouble with helping you out having such limited information. I'd need at least hub logs to get wider perspective. Are you a Samsung employee? It'd be easier to talk on Knox.

RaihaPark commented:
Yes, I work in Samsung. I'm happy to talk on Knox.

PKacprowiczS commented:
Great. Can you find me through my Knox ID, `p.kacprowicz`? Sadly, I wasn't able to find you.

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Update led-cpx-light.groovy

Deleted after "runLoaclly".

RaihaPark commented:
> Why not just include motion detection on the main DTH? Why break it out into a child?

Because customer want to see the motion detection and controls the on/off on the main card. I just follow customer's request.

greens commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,168 @@
+	
+	if (childDevice != null) {
+		childDevice.sendEvent(name: "motion", value: "inactive")
+	}
```
As @PKacprowiczS said, this will always be null if you call it here. I imagine you want to move this inside the `try` block

RaihaPark replied:
Yes, I changed it.

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Update led-cpx-light.groovy

Edited codes

RaihaPark commented:
Dear @PKacprowiczS @greens 
I change the location of line 156 into the try block. It works well. Please review again. :)

greens commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,167 @@
+ */
+
+metadata {
+	definition(name: "LED CPX light", namespace: "SAMSUNG LED", author: "SAMSUNG LED", runLocally: true) {
```
I don't think events will get properly dispatched to the child sensor if you have `runLocally` set to `true`

RaihaPark replied:
If I delete "runLocally: true", rather plug-in works strangely. It doesn't match to on/off status of device, plus fail to initializing.

RaihaPark commented:
@greens I really want to follow the code policy of ST. But it is very clear that if I deleted "runLocally: true", the plug-in works strangely. Log and phenomenon shows that. I'm afraid that if I deleted that code, WWST certification would be fail. @PKacprowiczS would you mind help me for explain, please? We were worked to aligning code and left "runLocally: true" inevitably.

PKacprowiczS commented:
Yeah, we've tried to figure it out together (as good as me changing DTH & looking at logs while @RaihaPark operates the device allows to do it), but I couldn't find out why it's not working correctly while executing in cloud. My understanding is that motion sensor events are correctly sent to child device, even when `runLocally` is set to `true`, right?

greens commented:
If the device only works when runLocally is true, that means there are bugs in the DTH. I'm not going to merge this knowing that the DTH code is faulty.

RaihaPark commented:
@greens If your opinion is right, zigbee-rgbw-bulb.groovy and zigbee-white-color-temperature-bulb.src is failed, too. If I delete after author zigbee-rgbw-bulb.groovy operates same and zigbee-white-color-temperature-bulb.src operate very strangely. zigbee-white-color-temperature-bulb.src is not works even though on/off. If I push button on/off then server error occurred. And it's not a device fault, it's only a plug-in error that plug-in on/off status changes automatically and initializing fail.

RaihaPark commented:
@greens Plus, sensor works well regardless runLocally. runLocally does not effect to the child event. It just effect to a parent plug-in not a device.

PKacprowiczS commented:
@RaihaPark those DTHes you've mentioned are not necessarily broken. Perhaps device sends _too much_ messages for each on/off where one of them has a wrong payload, and so just happens that local hub code handles only one of them, which has a correct status with it, while DTH code (or more likely `appengine-zigbee`) picks up every message, which leads to wrong status in the end.
You need to figure out which message (out of 3, as far as I can remember) has to be excluded from creating an event in the DTH.

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Update led-cpx-light.groovy

Delete "runlocally:true" and add a rule to the event in parser.

RaihaPark commented:
@PKacprowiczS @greens I edited the code to delete the "runlocally:true". The device and the plug-in works well. And there is some improvemnets in parse function. The location of variable zigbeeMap changed, and added a conditional statement for event. The conditional statement handles on/off clearly. I tested every event and function, there is no problem. Could you please review it again?

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -65,24 +65,29 @@ private getMOTION_CLUSTER() { 0x0406 }
 	if (event) {
+		if (zigbeeMap.clusterInt == ON_OFF_CLUSTER_VALUE && (zigbeeMap.data != ON_DATA_VALUE || zigbeeMap.data != OFF_DATA_VALUE)) {
+			return
+		}
```
Is there any particular reason that we need to check whole array? Wouldn't it work if only data[0] was checked?

RaihaPark replied:
Oh, I didn't thought that. OK, I'll change that.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -65,24 +65,29 @@ private getMOTION_CLUSTER() { 0x0406 }
 private getON_OFF_CLUSTER_VALUE() { 0x0006 }
 private getCONFIGURE_REPORTING_RESPONSE() { 0x07 }
+private getON_DATA_VALUE() { [0x01, 0x00] }
+private getOFF_DATA_VALUE() { [0x00, 0x00] }
```
Can we loose that `_VALUE` end? Just this time make sure that you're editing _every_ occurrence.

RaihaPark replied:
Sure, I changed those.

RaihaPark committed:
Update led-cpx-light.groovy

Aligned codes.

RaihaPark committed:
Update led-cpx-light.groovy

Edited codes.

PKacprowiczS commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,172 @@
+
+private getMOTION_CLUSTER() { 0x0406 }
+private getMOTION_STATUS_ATTRIBUTE() { 0x0000 }
+private getON_OFF_CLUSTER_VALUE() { 0x0006 }
```
Also this one to `getON_OFF_CLUSTER()` (and adjust appearance below)

RaihaPark replied:
Yes, thank you :)

RaihaPark committed:
Update led-cpx-light.groovy

Edited the magic number.

greens commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,172 @@
+				}
+				zigbeeMap.value = zigbeeMap.value.endsWith("01") ? "active" : "inactive"
+				zigbeeMap.name = "motion"
+				childDevice.sendEvent(zigbeeMap)
```
Doesn't `zigbeeMap` have a lot of extraneous values?

I think we should just be making a new map here.

RaihaPark replied:
Oh, I never expected about that. Yes, I made a zigbeeMap for child event.

RaihaPark committed:
Merge branch 'SmartThingsCommunity:master' into master

RaihaPark committed:
Update led-cpx-light.groovy

Added a separate zigbeeMap for child event : zigbeeMap_child in line91.

greens commented on devicetypes/smartthings/zigbee-motion-sensor-light.src/led-cpx-light.groovy:
```diff
@@ -0,0 +1,173 @@
+				sendEvent(name: "checkInterval", value: 60 * 12, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
+			}
+		} else {
+			def zigbeeMap_child = zigbee.parseDescriptionAsMap(description)
```
my suggestion was that you create an all new event 

`def event = createEvent(name: "motion", value: "active")`

`parseDescriptionAsMap` creates a populated map. You can see this because you're referencing `.clusterInt` on the next line.

RaihaPark replied:
OK, I got it! I fixed that.

RaihaPark committed:
Update led-cpx-light.groovy

Aligned send child event block.

RaihaPark commented:
@PKacprowiczS @greens Thanks for approval. And I'm sorry for all the troubles. I never had expected that "runlocally" means so important. Even if it was not so important, I should have to fixed that. That is really shame of me. It was my mistake. By the pressure of schedule, I think I was strange.

greens commented:
@RaihaPark It's okay. It's not really a well-documented feature because it's mostly intended for internal development.

samsung-iot-module commented:
@greens
I am scott kim working with youngoh park who is responsible for the DTH code of this project.
I have a problem with my device ICON on the smartthings app.
My device is a light, so I want to see the LIGHT BULB icon on my device card, but I could see the SWITCH icon now.
I want to change it to LIGHT BULB icon, could you let me know how to change it?
* I heard the reviewer told us to remove some metadata.definition code before, so we removed some code like [ocfDeviceType: "oic.d.light"] and so on,    
  Did this affect them?

KKlimczukS commented:
> @greens I am scott kim working with youngoh park who is responsible for the DTH code of this project. I have a problem with my device ICON on the smartthings app. My device is a light, so I want to see the LIGHT BULB icon on my device card, but I could see the SWITCH icon now. I want to change it to LIGHT BULB icon, could you let me know how to change it?
> 
> * I heard the reviewer told us to remove some metadata.definition code before, so we removed some code like [ocfDeviceType: "oic.d.light"] and so on,
>   Did this affect them?

@samsung-iot-module I would leave `ocfDeviceType: "oic.d.light"` in the metadata definition. Wasn't it working ?
(I see in this pull request that another value was used -> ocfDeviceType: "x.com.st.d.sensor.motion")
